packages/x-data-grid/src/hooks/features/rowSelection/utils.ts

```diff
@@ -13,3 +118,137 @@ export function isMultipleRowSelectionEnabled(
   }
   return !props.disableMultipleRowSelection;
 }
+
+const getRowNodeParents = (tree: GridRowTreeConfig, id: GridRowId) => {
+  const parents: GridRowId[] = [];
+
+  let parent: GridRowId | null = id;
+
+  while (isValidRowId(parent) && parent !== GRID_ROOT_GROUP_ID) {
+    const node = tree[parent] as GridGroupNode;
+    if (!node) {
+      return parents;
+    }
+    parents.push(parent);
+    parent = node.parent;
+  }
+  return parents;
+};
+
+const getFilteredRowNodeSiblings = (
+  tree: GridRowTreeConfig,
+  filteredRows: Record<GridRowId, boolean>,
+  id: GridRowId,
+) => {
+  const node = tree[id];
+  if (!node) {
+    return [];
+  }
+
+  const parent = node.parent;
+  if (!isValidRowId(parent)) {
+    return [];
+  }
+
+  const parentNode = tree[parent] as GridGroupNode;
+
+  return [...parentNode.children].filter((childId) => childId !== id && filteredRows[childId]);
```

Comment:
```suggestion
  return parentNode.children.filter((childId) => childId !== id && filteredRows[childId]);
```
`filter` returns a shallow copy anyway